# C14U5: Anthropology

In the following notebook, students will be able to:
* Create Python functions to process datasets using pandas
* Make a distinction between function definitions and function calls.
* Write code that accepts user input.



## 🟢 Module 1: Restaurant Culture Demographics Analysis


You will be given a dataset that provides restaurant inspections, violations, grades and adjudication information. Food is important in various cultures, and here we explore the types of food that each borough in NYC are exposed to.

First, use the pandas library to read in the data file.

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Display the first 5 rows of the dataset
df = pd.read_csv("/content/drive/MyDrive/restaurant_data_set_2.csv - restaurant_data_set_2.csv (1).csv")
df.head()

CAMIS  ... GRADE
0  41338311  ...   NaN
1  50055587  ...     A
2  50046026  ...   NaN
3  41576311  ...     A
4  50000641  ...   NaN

[5 rows x 9 columns]

In [ ]:
#Find how many rows the dataset contains.
len(df)

404249

**Review**: Use the .loc[] attribute to filter the dataset to only show the information for each borough of NYC. Store each borough's information in its own variable.

In [ ]:
#Brooklyn
df_brklyn = df.loc[df["BORO"] == "Brooklyn"]
df_brklyn.head()

CAMIS  ... GRADE
1   50055587  ...     A
9   50057082  ...     A
10  50035136  ...     A
13  41348754  ...     C
14  50008656  ...   NaN

[5 rows x 9 columns]

In [ ]:
#Queens
df_queens = df.loc[df["BORO"] == "Queens"]
df_queens.head()

CAMIS  ... GRADE
6   50052874  ...     B
17  41650198  ...     A
22  50033380  ...     A
24  41658324  ...     P
31  50063209  ...     A

[5 rows x 9 columns]

In [ ]:
#Manhattan
df_Manhattan = df.loc[df["BORO"] == "Manhattan"]
df_Manhattan.head()

CAMIS  ... GRADE
2  50046026  ...   NaN
3  41576311  ...     A
4  50000641  ...   NaN
5  41331752  ...     C
7  50099515  ...     B

[5 rows x 9 columns]

In [ ]:
#Bronx
df_Bronx = df.loc[df["BORO"] == "Bronx"]
df_Bronx.head()


CAMIS  ... GRADE
16   50002478  ...   NaN
18   50048064  ...   NaN
59   50039683  ...     B
109  50018680  ...     A
134  50003350  ...   NaN

[5 rows x 9 columns]

In [ ]:
#Staten Island
df_SI = df.loc[df["BORO"] == "Staten Island"]
df_SI.head()

CAMIS  ... GRADE
0   41338311  ...   NaN
20  50056878  ...     A
21  50059651  ...   NaN
25  50044877  ...     B
26  40996642  ...     C

[5 rows x 9 columns]

### Investigate
Fast food is a convenience such that you can grab it and go quickly on a work day and it is cheap. The downside is that it is unhealthy. Yet somehow many low-income and food-insecure households rely on fast-food restaurants. 

Write a program that finds the percentage of each of the following fast-food restaurants in each borough.
* Mcdonald's 
* Burger King
* Wendy's 
* Popeyes

In [ ]:
#O'l Mcdonalds

#Brooklyn
len(df_brklyn.loc[df["DBA"] == "MCDONALD'S"])
#total of the restaurants:
total = len(df.loc[df["DBA"] == "MCDONALD'S"])
#find teh percent
print(len(df_brklyn.loc[df["DBA"] == "MCDONALD'S"])/total * 100, "%")


#Queenz
print(len(df_queens.loc[df["DBA"] == "MCDONALD'S"])/total * 100, "%")

#Manhattan
print(len(df_Manhattan.loc[df["DBA"] == "MCDONALD'S"])/total * 100, "%")

#Bronx
print(len(df_Bronx.loc[df["DBA"] == "MCDONALD'S"])/total * 100, "%")

#Staten Island
print(len(df_SI.loc[df["DBA"] == "MCDONALD'S"])/total * 100, "%")

20.269642026964203 %
25.615992561599256 %
29.242212924221295 %
20.82752208275221 %
4.044630404463041 %


In [ ]:
#Burger King


#brooklyn

#queens

#manhattan

#Bronx

#Staten Island

Change your code into a function that can produce percentages based on what restaurant you are searching the dataframe for. 

In [ ]:
#function Header
def borough_restaurantz_percentage(dataFrame, restaurant):
  #Brooklyn
  #total of the restaurants:
  total = len(dataFrame.loc[dataFrame["DBA"] == restaurant])
  #find teh percent
  print(len(dataFrame.loc[dataFrame["BORO"] == "Brooklyn"].loc[dataFrame["DBA"] == restaurant])/total * 100, "%")


  #Queenz
  print(len(dataFrame.loc[dataFrame["BORO"] == "Queens"].loc[dataFrame["DBA"] == restaurant])/total * 100, "%")

  #Manhattan
  print(len(dataFrame.loc[dataFrame["BORO"] == "Manhattan"].loc[dataFrame["DBA"] == restaurant])/total * 100, "%")

  #Bronx
  print(len(dataFrame.loc[dataFrame["BORO"] == "Bronx"].loc[dataFrame["DBA"] == restaurant])/total * 100, "%")

  #Staten Island
  print(len(dataFrame.loc[dataFrame["BORO"] == "Staten Island"].loc[dataFrame["DBA"] == restaurant])/total * 100, "%")

  return

In [ ]:
#Don't forget to call your function!
borough_restaurantz_percentage(df, "MCDONALD'S")

20.269642026964203 %
25.615992561599256 %
29.242212924221295 %
20.82752208275221 %
4.044630404463041 %


In [ ]:
borough_restaurantz_percentage(df, "BURGER KING")

36.78474114441417 %
26.430517711171664 %
11.353315168029065 %
19.346049046321525 %
6.085376930063579 %


In [ ]:
borough_restaurantz_percentage(df, "WENDY'S")

21.153846153846153 %
24.57264957264957 %
29.700854700854702 %
17.307692307692307 %
7.264957264957266 %


In [ ]:
borough_restaurantz_percentage(df, "POPEYES")

26.01787487586892 %
27.110228401191662 %
18.669314796425024 %
25.02482621648461 %
3.1777557100297913 %


Each borough has the following demographics (taken from US census data): <br><br>

Brooklyn
* White - 49.8%
* Black - 33.8%
* Native American - 0.9% 
* Asian - 12.7%
* Native Hawaiian and other Pacific Islander - 0.1%
* Mixed - 2.7%
* Hispanic or Latino - 18.9% 
<br><br>

Queens
* White - 47.8%
* Black - 20.7%
* Native American - 1.3% 
* Asian - 26.9%
* Native Hawaiian and other Pacific Islander - 0.2%
* Mixed - 3.0%
* Hispanic or Latino - 28.2%
<br><br>

Manhattan
* White - 64.6%
* Black - 17.8%
* Native American - 1.2%
* Asian - 12.8%
* Native Hawaiian and other Pacific Islander - 0.2% 
* Mixed - 3.4%
* Hispanic or Latino - 25.6% 
<br><br>

Bronx
* White -  44.7%
* Black - 24.3%
* Native American - 0.4% 
* Asian - 14.1%
* Native Hawaiian and other Pacific Islander - 0.1% 
* Mixed - 3.6%
* Hispanic or Latino - 29.1% 
<br><br>

Staten Island
* White - 74.5%
* Black - 11.6%
* Native American - 0.7% 
* Asian - 10.9%
* Native Hawaiian and other Pacific Islander - 0.1% 
* Mixed - 2.2%
* Hispanic or Latino - 18.6% 


Questions:
1. How can we narrow down the data further to find how different racial/ethnic groups are impacted by its population of fast-food restaurants? (by zip code)

2. How may the connection between fast-food restaurants and black and low-income neighborhoods contribute to the causes of the obesity epidemic in these poulations?

3. What can we do with computer science to help change this?

# 🟢 Module 2: Language Translation Function

In 1971, a linguist named Morris Swadesh published a list of words most resistant to cross-cultural change. The list, now referred to as the Swadesh list (there are two primary versions, of 100 and 200 words), has been used to reconstruct genealogical relationships among the languages in a language family.

AMNH colleagues, Dr. Whiteley and Dr. Ward Wheeler (Curator, Invertebrate Zoology) have been studying historical relationships among languages, beginning with Uto-Aztecan languages of Middle and North America, using genetic methods developed in the field of evolutionary biology. With phylogenetic software, they systematically compare cognate words as sequences of sounds (phonemes). Where available, they also analyze grammatical (morphological and syntactic) data.

**TASK**: Use the data given to write a program that accepts user-input and translates a given English word using the Uto-Aztecan language. Note: Not every English word is recorded, so be sure to create a conditional for the case that it does not exist in the dataset.

In [ ]:
import pandas as pd
df_languages = pd.read_csv("/content/drive/MyDrive/uto_aztecan_languages_datasets.csv")
df_languages.head()

Language  ...                 Unnamed: 204
0                               NaN  ...                         tipa
1                               NaN  ...                          NaN
2                   Northern_Paiute  ...               t\=Wt\super si
3                      Western_Mono  ...   inott\super sPtt\super siP
4  Tümpisa_Shoshone_(Panamint,Koso)  ...  tWtWtt\super sitt\super si 

[5 rows x 205 columns]

In [ ]:
for i in df_languages.columns:
  if "Unnamed" in i:
    dflanguages = df_languages.drop([i], axis = 1)

print(df_languages)

                            Language  ...                                       Unnamed: 204
0                                NaN  ...                                               tipa
1                                NaN  ...                                                NaN
2                    Northern_Paiute  ...                                     t\=Wt\super si
3                       Western_Mono  ...                         inott\super sPtt\super siP
4   Tümpisa_Shoshone_(Panamint,Koso)  ...                        tWtWtt\super sitt\super si 
5         Big_Smokey_Valley_Shoshone  ...                           tW{\textipa{-}} | tWtWn 
6                   Western_Shoshone  ...                                             tWiku 
7                           Shoshone  ...                                               tWeh
8                           Comanche  ...                                          tWt\=AttW
9                           Kawaiisu  ...  \=AkkAn\=utt\super Si | \=W

# 🟢 Module 3: Error Rates in Users of Automatic Face Recognition Software

While facial recognition sounds nice and may have its perks, there is bias in all algorithms as long as there is a lack of diversity in its creators. If not made correctly, facial recognition risks being weaponized by law enforcement against marginalized communities around the world. From New Delhi to New York, this invasive technology can turn our identities against us and can undermine human rights. 

Use the given dataset to create a plot of the error rates based on gender and race. You will be given an excel file and are responsible for saving it as a csv file to use.